x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -116,9 +116,20 @@ public void tryAction(ActionListener<InferenceServiceResults> listener) {
             try {
                 if (request.isStreaming() && responseHandler.canHandleStreamingResponses()) {
                     httpClient.stream(request.createHttpRequest(), context, retryableListener.delegateFailure((l, r) -> {
-                        var streamingResponseHandler = new StreamingResponseHandler(throttlerManager, logger, request, responseHandler);
-                        r.subscribe(streamingResponseHandler);
-                        l.onResponse(responseHandler.parseResult(request, streamingResponseHandler));
+                        if (r.isSuccessfulResponse()) {
+                            l.onResponse(responseHandler.parseResult(request, r.toHttpResult()));
+                        } else {
+                            r.readFullResponse(l.delegateFailureAndWrap((ll, httpResult) -> {
+                                try {
+                                    responseHandler.validateResponse(throttlerManager, logger, request, httpResult);
+                                    InferenceServiceResults inferenceResults = responseHandler.parseResult(request, httpResult);
+                                    ll.onResponse(inferenceResults);
```

Comment:
Just to make sure I understand this flow correctly, we can get a status code that indicates a failure but if `validateResponse` doesn't throw an error we'll return an actual result? Or are we calling `onResponse` here to also handle returning an error object?